clang/test/CXX/drs/dr201.cpp

```diff
@@ -0,0 +1,43 @@
+// RUN: %clang_cc1 -std=c++98 %s -triple x86_64-linux-gnu -emit-llvm -disable-llvm-passes -O0 -o - -fexceptions -fcxx-exceptions -pedantic-errors | llvm-cxxfilt -n | FileCheck %s --check-prefixes CHECK
+// RUN: %clang_cc1 -std=c++11 %s -triple x86_64-linux-gnu -emit-llvm -disable-llvm-passes -O0 -o - -fexceptions -fcxx-exceptions -pedantic-errors | llvm-cxxfilt -n | FileCheck %s --check-prefixes CHECK
+// RUN: %clang_cc1 -std=c++14 %s -triple x86_64-linux-gnu -emit-llvm -disable-llvm-passes -O0 -o - -fexceptions -fcxx-exceptions -pedantic-errors | llvm-cxxfilt -n | FileCheck %s --check-prefixes CHECK
+// RUN: %clang_cc1 -std=c++17 %s -triple x86_64-linux-gnu -emit-llvm -disable-llvm-passes -O0 -o - -fexceptions -fcxx-exceptions -pedantic-errors | llvm-cxxfilt -n | FileCheck %s --check-prefixes CHECK
+// RUN: %clang_cc1 -std=c++20 %s -triple x86_64-linux-gnu -emit-llvm -disable-llvm-passes -O0 -o - -fexceptions -fcxx-exceptions -pedantic-errors | llvm-cxxfilt -n | FileCheck %s --check-prefixes CHECK
+// RUN: %clang_cc1 -std=c++23 %s -triple x86_64-linux-gnu -emit-llvm -disable-llvm-passes -O0 -o - -fexceptions -fcxx-exceptions -pedantic-errors | llvm-cxxfilt -n | FileCheck %s --check-prefixes CHECK
+// RUN: %clang_cc1 -std=c++2c %s -triple x86_64-linux-gnu -emit-llvm -disable-llvm-passes -O0 -o - -fexceptions -fcxx-exceptions -pedantic-errors | llvm-cxxfilt -n | FileCheck %s --check-prefixes CHECK
+
+// RUN: %clang_cc1 -std=c++98 %s -triple x86_64-linux-gnu -emit-llvm -disable-llvm-passes -O3 -o - -fexceptions -fcxx-exceptions -pedantic-errors | llvm-cxxfilt -n | FileCheck %s --check-prefixes CHECK
+// RUN: %clang_cc1 -std=c++11 %s -triple x86_64-linux-gnu -emit-llvm -disable-llvm-passes -O3 -o - -fexceptions -fcxx-exceptions -pedantic-errors | llvm-cxxfilt -n | FileCheck %s --check-prefixes CHECK
+// RUN: %clang_cc1 -std=c++14 %s -triple x86_64-linux-gnu -emit-llvm -disable-llvm-passes -O3 -o - -fexceptions -fcxx-exceptions -pedantic-errors | llvm-cxxfilt -n | FileCheck %s --check-prefixes CHECK
+// RUN: %clang_cc1 -std=c++17 %s -triple x86_64-linux-gnu -emit-llvm -disable-llvm-passes -O3 -o - -fexceptions -fcxx-exceptions -pedantic-errors | llvm-cxxfilt -n | FileCheck %s --check-prefixes CHECK
+// RUN: %clang_cc1 -std=c++20 %s -triple x86_64-linux-gnu -emit-llvm -disable-llvm-passes -O3 -o - -fexceptions -fcxx-exceptions -pedantic-errors | llvm-cxxfilt -n | FileCheck %s --check-prefixes CHECK
+// RUN: %clang_cc1 -std=c++23 %s -triple x86_64-linux-gnu -emit-llvm -disable-llvm-passes -O3 -o - -fexceptions -fcxx-exceptions -pedantic-errors | llvm-cxxfilt -n | FileCheck %s --check-prefixes CHECK
+// RUN: %clang_cc1 -std=c++2c %s -triple x86_64-linux-gnu -emit-llvm -disable-llvm-passes -O3 -o - -fexceptions -fcxx-exceptions -pedantic-errors | llvm-cxxfilt -n | FileCheck %s --check-prefixes CHECK
```

Comment:
Unfortunately, CodeGen tests cover both language frontend and codegen, so we have 7 major frontend codepaths times 2 major codegen codepaths to test.

Reply:
We do not write -O3 tests because that's a concern for backend folks. 
And generally, unless there have been significant changes there is no value in testing all standard versions

Reply:
> We do not write -O3 tests because that's a concern for backend folks.

Clang's CodeGen is also concerned whether optimization are enabled or not in 55 places. I don't want to miss those codepaths.

> And generally, unless there have been significant changes there is no value in testing all standard versions

Those are not general tests, but defect report tests. I believe they should be held to higher standards.

Reply:
I think a good litmus test as to whether a test belong in Clang or not is, "if the test fail, can we fix it?"
And here we would not be able to, lowering IR is not in the purview of the front end, so this is not testing the front end.

And sure, there is a wider question here: Does the LLVM middle-end implements and test the semantics of C++
and those core issues that are relevant?

But regardless of the answer that question, putting such test here would not serve to increase out coverage.
We test that clang produces good IR, and we trust that given the IR of a well-formed program LLVM will preserve the semantics.

Ditto we should not test libc++, libc++abi, etc

Reply:
> Clang's CodeGen is also concerned whether optimization are enabled or not in 55 places. I don't want to miss those codepaths.

Keep in mind that changes get tested in multiple places (when done correctly), so when someone implements a DR (or a feature with the DR already applied), they should have test coverage in `clang/test/CodeGen*` as well as in `clang/test/Sema*`, and here in the DR test file. We don't need to re-test things covered by other tests, and sometimes the coverage makes more sense elsewhere than a DR test.

I tend to agree with @cor3ntin that this may be going overboard in terms of test runs. I would drop the -O3 tests (and remove the -O0 from the other tests) unless there's some evidence that we're increasing test coverage we don't otherwise get. (We have a coverage report (https://lab.llvm.org/coverage/coverage-reports/coverage/Users/buildslave/jenkins/workspace/coverage/llvm-project/index.html) but it doesn't get run regularly, but that would be one way to show that the extra test coverage is testing more paths.)

Reply:
> And sure, there is a wider question here: Does the LLVM middle-end implements and test the semantics of C++ and those core issues that are relevant?

CC @lattner as [general LLVM code owner](https://github.com/llvm/llvm-project/blob/26db3c3b72d3c915ad296a5a5313210bde8ce3e1/llvm/CODE_OWNERS.TXT#L154), both for an answer to this question, but also as a request to find a dedicated code owner within LLVM that the Clang community can interface with regarding C and C++ conformance from the LLVM side of things. (For example, we really need someone on the LLVM side who we can talk to regarding changes to the memory object model in both C and C++ as the languages evolve so that we can plan a path forward accordingly.)

Reply:
> I think a good litmus test as to whether a test belong in Clang or not is, "if the test fail, can we fix it?"
And here we would not be able to, lowering IR is not in the purview of the front end, so this is not testing the front end.

Lowering AST to IR _is_ in purview of Clang, and we're responsible that this happens correctly (even if no single individual in this conversation is an expert in that part of Clang). I hope everyone here agrees that middle-end folks are not responsible for the correctness of our CodeGen from the C++ language semantics perspective.

> And sure, there is a wider question here: Does the LLVM middle-end implements and test the semantics of C++
and those core issues that are relevant?

No, there is no such question, and there wasn't at any point in DR testing discussion. CodeGen DR tests are interested in Clang's CodeGen, and I do my best to ensure that middle-end does not interfere. If you think middle-end is still involved somehow, let me know how this can be fixed.

> Keep in mind that changes get tested in multiple places (when done correctly), so when someone implements a DR (or a feature with the DR already applied), they should have test coverage in clang/test/CodeGen* as well as in clang/test/Sema*, and here in the DR test file. We don't need to re-test things covered by other tests, and sometimes the coverage makes more sense elsewhere than a DR test.

This holds for the "usual" DR tests as well. They sure duplicate some of Sema tests. Same goes for the whole CXX directory. On top of the fact that calculating coverage is both time- and resource- consuming, making automatic duplicate discovery a serious burden, regular Sema and CodeGen tests are not necessarily designed to cover use cases DR tests cover. Which means they can change in a ways not fit for DR testing. So we should be careful about depending on them.

> I tend to agree with @cor3ntin that this may be going overboard in terms of test runs.

It wouldn't take too much time for a list of language modes to have the same length. Is cutting a number of language modes DR tests are run in also on the table because of arbitrary limits (as I see them at the moment)?

Reply:
CC @rjmccall and @efriedma-quic as Clang codegen code owners.

> Lowering AST to IR _is_ in purview of Clang, and we're responsible that this happens correctly (even if no single individual in this conversation is an expert in that part of Clang). I hope everyone here agrees that middle-end folks are not responsible for the correctness of our CodeGen from the C++ language semantics perspective.

There's a handshake involved between Clang and LLVM, so I don't think there's *one* responsible party for correctness here. Part of the correctness comes from Clang lowering to sensible LLVM IR, but we sometimes need LLVM folks to tell us what that sensible IR is or to provide different tools for us if existing IR doesn't suffice. However, in terms of DR testing, I think most of the effort should be expended on Clang's side of things.

> It wouldn't take too much time for a list of language modes to have the same length. Is cutting a number of language modes DR tests are run in also on the table because of arbitrary limits (as I see them at the moment)?

IMO, we want to find a happy middle ground between testing too much and testing too little. The issues with testing too little are obvious, but the issue from over-testing is that we waste significant resources for little benefit. Everyone runs the test suite (precommit CI, postcommit CI, and developers on their local machines) so if we increase the test time significantly without discernible benefit, that can lead to death-by-a-thousand-cuts problems.

FWIW, I wasn't insisting on code coverage measurements. I was more pointing out that I think we should not add the extra RUN lines for -O3 while speculating that it increases test coverage. We should have some reason to believe it increases test coverage, and we might want to consider splitting those tests out into their own file so we don't run 100 DR tests in two optimization modes for every standard version we're testing.

Reply:
> We should have some reason to believe it increases test coverage, and we might want to consider splitting those tests out into their own file so we don't run 100 DR tests in two optimization modes for every standard version we're testing.

But... every codegen tests is in its own file already because of an entirely different concern. I don't see us being able to combine them into a single file for foreseeable future. So running 100 DRs with those additional flags is not on the table.

Reply:
"DR tests" are not subject to a higher standard than ordinary tests.  Many ordinary tests are functionality tests for basic specified behavior and are not somehow of lesser significance.  This sort of exhaustive testing of configurations would be reasonable for an external conformance test suite, in which case the general structure of the test suite would run every test in every different configuration and the individual tests would just note expected differences by configuration.  However, the Clang regression test suite is not a conformance test suite; it is a set of targeted functionality tests which is designed to detect bugs, regressions, and other behavior changes with tolerably low overhead, given that it is repeatedly run in CI and individual user testing.  Tests should therefore be written with reasonable human knowledge rather than using brute force. You can and should test as many configurations in a test as your knowledge of the code suggests that it is likely that the compiler will take different code paths and/or generate different code patterns for.